designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,459 @@
+
+![ESLint](ESLint.png)
+
+- When rules are suppressed in-source (`eslint-disable` inside of a file), ESLint will still run the rules but then remove any lint messages that were triggered with the specified rule(s).
+
+### What we need
+
+As mentioned above, what we need is to allow engineers to provide justification in the suppression and output the suppression information.
+
+To minimize changes, we propose to use the `-- Description` part of the suppression as "Justification". More specifically, for directive suppression (or inline directive comments in ESLint), the description after two or more consecutive dashes can be treated as the justification.
+
+For example:
+
+```javascript
+/* eslint-disable no-undef -- Justification example */
+a = 1;
+/* eslint-enable no-undef */
+```
+```javascript
+console.log("foo"); // eslint-disable-line no-console -- Justification example
+```
+```javascript
+// eslint-disable-next-line no-console -- Justification example
+console.log("bar");
+```
+
+Then we may get:
+
+```json
+{"kind": "directive", "justification": "Justification example"}
+```
+
+If a violation is inline-suppressed multiple times, all suppression entries will be recorded, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852). Meanwhile, only the first suppression will be tagged as "used" as ESLint does currently, and the rest of the suppressions should be in `unusedDisableDirectives` so that the behavior of `--report-unused-disable-directives` will not be changed. A suppression list would be preferred to gather all suppression information of a violation:
+
+```json
+suppressions: [
+  {"kind": "directive", "justification": "Justification message 1"},
+  {"kind": "directive", "justification": "Justification message 2"}
+]
+```
+
+## Detailed Design
+
+### Goal
+
+1. Keep the current behavior of ESLint, i.e., errors/warnings will not be shown on the command line if the related rules are disabled, unless `--no-inline-config` is used.
+2. The violations (problems) that disabled by inline comments will not be removed.
+3. Add suppression info in the output of ESLint. For directive suppressions, we generate suppression info where `kind` is `directive` and `justification` is the description in the directive comments (words after two or more consecutive dashes mentioned above).
+    
+    For example:
+        
+    ```javascript
+    /* eslint-disable no-undef -- Justification example */
+    a = 1;
+    /* eslint-enable no-undef */
+    ```
+    ```javascript
+    console.log("foo"); // eslint-disable-line no-console -- Justification example
+    ```
+    ```javascript
+    // eslint-disable-next-line no-console -- Justification example
+    console.log("bar");
+    ```
+
+    Output:
+
+    ```json
+    {"kind": "directive", "justification": "Justification example"}
+    ```
+
+### General Design
```

Comment:
Can you describe in this section changes that would be made on the [Linter API](https://eslint.org/docs/developer-guide/nodejs-api#linter)? How will `Linter` return suppressed messages, and should we add some new options there? Note that this is a public API, and all changes on it should be backward compatible.

Reply:
Updated. Is the description good to you?

Reply:
Thanks! I left a question there.